File: examples/api_client.py
#!/usr/bin/env python3
"""
Cliente de ejemplo para la API de Análisis de Presentaciones Empresariales
"""

import requests
import json
import os
from typing import Dict, Any, Optional

class PresentationAnalysisAPI:
    """Cliente para la API de análisis de presentaciones"""
    
    def __init__(self, base_url: str, api_key: str):
        """
        Inicializar el cliente API
        
        Args:
            base_url: URL base de la API (ej: https://tu-dominio.onrender.com/api/v1)
            api_key: API key para autenticación
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.headers = {
            'X-API-Key': api_key,
            'Content-Type': 'application/json'
        }
    
    def health_check(self) -> Dict[str, Any]:
        """Verificar estado de la API"""
        response = requests.get(f"{self.base_url}/health")
        return response.json()
    
    def upload_presentation(self, 
                          file_path: str, 
                          rfc: str, 
                          company_name: str = "", 
                          presenter_name: str = "") -> Dict[str, Any]:
        """
        Subir y analizar una presentación
        
        Args:
            file_path: Ruta al archivo de video
            rfc: RFC de la empresa
            company_name: Nombre de la empresa (opcional)
            presenter_name: Nombre del presentador (opcional)
        
        Returns:
            Respuesta de la API con el análisis
        """
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"Archivo no encontrado: {file_path}")
        
        url = f"{self.base_url}/presentations/upload"
        
        with open(file_path, 'rb') as f:
            files = {'file': f}
            data = {
                'rfc': rfc,
                'company_name': company_name,
                'presenter_name': presenter_name
            }
            headers = {'X-API-Key': self.api_key}
            
            response = requests.post(url, headers=headers, files=files, data=data)
            
            if response.status_code != 200:
                raise Exception(f"Error en la API: {response.status_code} - {response.text}")
            
            return response.json()
    
    def get_presentation(self, rfc: str) -> Dict[str, Any]:
        """
        Obtener análisis de una presentación
        
        Args:
            rfc: RFC de la empresa
        
        Returns:
            Datos de la presentación y análisis
        """
        url = f"{self.base_url}/presentations/{rfc}"
        response = requests.get(url, headers=self.headers)
        
        if response.status_code == 404:
            raise Exception(f"Presentación no encontrada para RFC: {rfc}")
        
        if response.status_code != 200:
            raise Exception(f"Error en la API: {response.status_code} - {response.text}")
        
        return response.json()
    
    def get_presentation_status(self, rfc: str) -> Dict[str, Any]:
        """
        Verificar estado de procesamiento
        
        Args:
            rfc: RFC de la empresa
        
        Returns:
            Estado de la presentación
        """
        url = f"{self.base_url}/presentations/{rfc}/status"
        response = requests.get(url, headers=self.headers)
        return response.json()
    
    def list_presentations(self, limit: int = 50, offset: int = 0) -> Dict[str, Any]:
        """
        Listar presentaciones
        
        Args:
            limit: Número máximo de resultados
            offset: Offset para paginación
        
        Returns:
            Lista de presentaciones
        """
        url = f"{self.base_url}/presentations"
        params = {'limit': limit, 'offset': offset}
        response = requests.get(url, headers=self.headers, params=params)
        
        if response.status_code != 200:
            raise Exception(f"Error en la API: {response.status_code} - {response.text}")
        
        return response.json()
    
    def analyze_transcript(self, transcript: str) -> Dict[str, Any]:
        """
        Analizar solo un transcript
        
        Args:
            transcript: Texto a analizar
        
        Returns:
            Análisis del transcript
        """
        url = f"{self.base_url}/analysis/transcript"
        data = {'transcript': transcript}
        response = requests.post(url, headers=self.headers, json=data)
        
        if response.status_code != 200:
            raise Exception(f"Error en la API: {response.status_code} - {response.text}")
        
        return response.json()
    
    def get_metrics(self) -> Dict[str, Any]:
        """Obtener métricas de la API"""
        url = f"{self.base_url}/metrics"
        response = requests.get(url, headers=self.headers)
        
        if response.status_code != 200:
            raise Exception(f"Error en la API: {response.status_code} - {response.text}")
        
        return response.json()


def main():
    """Ejemplo de uso del cliente API"""
    
    # Configuración
    BASE_URL = "https://tu-dominio.onrender.com/api/v1"  # Cambiar por tu URL
    API_KEY = "tu-api-key-aqui"  # Cambiar por tu API key
    
    # Crear cliente
    api = PresentationAnalysisAPI(BASE_URL, API_KEY)
    
    try:
        # 1. Verificar estado de la API
        print("=== Verificando estado de la API ===")
        health = api.health_check()
        print(f"Estado: {health['status']}")
        print(f"Versión: {health['version']}")
        print()
        
        # 2. Subir y analizar una presentación
        print("=== Subiendo presentación ===")
        # Descomenta las siguientes líneas para subir un archivo real
        # result = api.upload_presentation(
        #     file_path="presentacion.mp4",
        #     rfc="ABC123456789",
        #     company_name="Mi Empresa S.A.",
        #     presenter_name="Juan Pérez"
        # )
        # print(f"Análisis completado: {result['message']}")
        # print(f"Puntuación general: {result['analysis']['scores']['overall']}")
        # print()
        
        # 3. Analizar un transcript de ejemplo
        print("=== Analizando transcript ===")
        transcript = """
        Buenos días, hoy les voy a presentar nuestra empresa Tecnología Avanzada S.A.
        Somos una empresa fundada en 2010 que se dedica al desarrollo de software empresarial.
        Nuestra misión es facilitar la transformación digital de las empresas mexicanas.
        Ofrecemos soluciones en gestión de recursos empresariales, CRM y automatización.
        Nuestros valores principales son la innovación, la calidad y el servicio al cliente.
        Hemos logrado implementar más de 500 proyectos exitosos en todo el país.
        """
        
        analysis = api.analyze_transcript(transcript)
        print("Análisis del transcript:")
        print(f"Puntuación general: {analysis['analysis']['scores']['overall']}")
        print(f"Resumen: {analysis['analysis']['resumen']}")
        print("Feedback:")
        for feedback in analysis['analysis']['feedback']:
            print(f"  - {feedback}")
        print()
        
        # 4. Obtener métricas
        print("=== Métricas de la API ===")
        metrics = api.get_metrics()
        print(f"Total de presentaciones: {metrics['total_presentations']}")
        print(f"Tamaño total de almacenamiento: {metrics['total_storage_size_mb']} MB")
        print()
        
        # 5. Listar presentaciones (si hay alguna)
        print("=== Listando presentaciones ===")
        presentations = api.list_presentations(limit=5)
        print(f"Total de presentaciones encontradas: {presentations['total']}")
        for presentation in presentations['presentations']:
            print(f"  - RFC: {presentation['rfc']}")
            print(f"    Empresa: {presentation.get('company_name', 'N/A')}")
            print(f"    Fecha: {presentation.get('upload_timestamp', 'N/A')}")
            print()
        
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main() 

File: examples/api_client.js
/**
 * Cliente de ejemplo para la API de Análisis de Presentaciones Empresariales
 * Versión JavaScript para uso en navegador o Node.js
 */

class PresentationAnalysisAPI {
    /**
     * Constructor del cliente API
     * @param {string} baseUrl - URL base de la API
     * @param {string} apiKey - API key para autenticación
     */
    constructor(baseUrl, apiKey) {
        this.baseUrl = baseUrl.replace(/\/$/, ''); // Remover trailing slash
        this.apiKey = apiKey;
        this.headers = {
            'X-API-Key': apiKey,
            'Content-Type': 'application/json'
        };
    }

    /**
     * Verificar estado de la API
     * @returns {Promise<Object>} Estado de la API
     */
    async healthCheck() {
        const response = await fetch(`${this.baseUrl}/health`);
        return await response.json();
    }

    /**
     * Subir y analizar una presentación
     * @param {File} file - Archivo de video
     * @param {string} rfc - RFC de la empresa
     * @param {string} companyName - Nombre de la empresa (opcional)
     * @param {string} presenterName - Nombre del presentador (opcional)
     * @returns {Promise<Object>} Resultado del análisis
     */
    async uploadPresentation(file, rfc, companyName = '', presenterName = '') {
        const formData = new FormData();
        formData.append('file', file);
        formData.append('rfc', rfc);
        formData.append('company_name', companyName);
        formData.append('presenter_name', presenterName);

        const response = await fetch(`${this.baseUrl}/presentations/upload`, {
            method: 'POST',
            headers: {
                'X-API-Key': this.apiKey
            },
            body: formData
        });

        if (!response.ok) {
            throw new Error(`Error en la API: ${response.status} - ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Obtener análisis de una presentación
     * @param {string} rfc - RFC de la empresa
     * @returns {Promise<Object>} Datos de la presentación
     */
    async getPresentation(rfc) {
        const response = await fetch(`${this.baseUrl}/presentations/${rfc}`, {
            headers: this.headers
        });

        if (response.status === 404) {
            throw new Error(`Presentación no encontrada para RFC: ${rfc}`);
        }

        if (!response.ok) {
            throw new Error(`Error en la API: ${response.status} - ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Verificar estado de procesamiento
     * @param {string} rfc - RFC de la empresa
     * @returns {Promise<Object>} Estado de la presentación
     */
    async getPresentationStatus(rfc) {
        const response = await fetch(`${this.baseUrl}/presentations/${rfc}/status`, {
            headers: this.headers
        });
        return await response.json();
    }

    /**
     * Listar presentaciones
     * @param {number} limit - Número máximo de resultados
     * @param {number} offset - Offset para paginación
     * @returns {Promise<Object>} Lista de presentaciones
     */
    async listPresentations(limit = 50, offset = 0) {
        const params = new URLSearchParams({
            limit: limit.toString(),
            offset: offset.toString()
        });

        const response = await fetch(`${this.baseUrl}/presentations?${params}`, {
            headers: this.headers
        });

        if (!response.ok) {
            throw new Error(`Error en la API: ${response.status} - ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Analizar solo un transcript
     * @param {string} transcript - Texto a analizar
     * @returns {Promise<Object>} Análisis del transcript
     */
    async analyzeTranscript(transcript) {
        const response = await fetch(`${this.baseUrl}/analysis/transcript`, {
            method: 'POST',
            headers: this.headers,
            body: JSON.stringify({ transcript })
        });

        if (!response.ok) {
            throw new Error(`Error en la API: ${response.status} - ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Obtener métricas de la API
     * @returns {Promise<Object>} Métricas de la API
     */
    async getMetrics() {
        const response = await fetch(`${this.baseUrl}/metrics`, {
            headers: this.headers
        });

        if (!response.ok) {
            throw new Error(`Error en la API: ${response.status} - ${response.statusText}`);
        }

        return await response.json();
    }
}

// Ejemplo de uso en navegador
async function exampleUsage() {
    // Configuración
    const BASE_URL = 'https://tu-dominio.onrender.com/api/v1'; // Cambiar por tu URL
    const API_KEY = 'tu-api-key-aqui'; // Cambiar por tu API key

    // Crear cliente
    const api = new PresentationAnalysisAPI(BASE_URL, API_KEY);

    try {
        // 1. Verificar estado de la API
        console.log('=== Verificando estado de la API ===');
        const health = await api.healthCheck();
        console.log('Estado:', health.status);
        console.log('Versión:', health.version);
        console.log();

        // 2. Analizar un transcript de ejemplo
        console.log('=== Analizando transcript ===');
        const transcript = `
            Buenos días, hoy les voy a presentar nuestra empresa Tecnología Avanzada S.A.
            Somos una empresa fundada en 2010 que se dedica al desarrollo de software empresarial.
            Nuestra misión es facilitar la transformación digital de las empresas mexicanas.
            Ofrecemos soluciones en gestión de recursos empresariales, CRM y automatización.
            Nuestros valores principales son la innovación, la calidad y el servicio al cliente.
            Hemos logrado implementar más de 500 proyectos exitosos en todo el país.
        `;

        const analysis = await api.analyzeTranscript(transcript);
        console.log('Análisis del transcript:');
        console.log('Puntuación general:', analysis.analysis.scores.overall);
        console.log('Resumen:', analysis.analysis.resumen);
        console.log('Feedback:');
        analysis.analysis.feedback.forEach(feedback => {
            console.log('  -', feedback);
        });
        console.log();

        // 3. Obtener métricas
        console.log('=== Métricas de la API ===');
        const metrics = await api.getMetrics();
        console.log('Total de presentaciones:', metrics.total_presentations);
        console.log('Tamaño total de almacenamiento:', metrics.total_storage_size_mb, 'MB');
        console.log();

        // 4. Listar presentaciones
        console.log('=== Listando presentaciones ===');
        const presentations = await api.listPresentations(5);
        console.log('Total de presentaciones encontradas:', presentations.total);
        presentations.presentations.forEach(presentation => {
            console.log('  - RFC:', presentation.rfc);
            console.log('    Empresa:', presentation.company_name || 'N/A');
            console.log('    Fecha:', presentation.upload_timestamp || 'N/A');
            console.log();
        });

    } catch (error) {
        console.error('Error:', error.message);
    }
}

// Ejemplo de uso con subida de archivo (para navegador)
async function uploadFileExample() {
    const BASE_URL = 'https://tu-dominio.onrender.com/api/v1';
    const API_KEY = 'tu-api-key-aqui';
    const api = new PresentationAnalysisAPI(BASE_URL, API_KEY);

    // Obtener archivo del input file
    const fileInput = document.getElementById('fileInput');
    const file = fileInput.files[0];
    
    if (!file) {
        alert('Por favor selecciona un archivo');
        return;
    }

    try {
        console.log('Subiendo archivo:', file.name);
        
        const result = await api.uploadPresentation(
            file,
            'ABC123456789',
            'Mi Empresa S.A.',
            'Juan Pérez'
        );

        console.log('Análisis completado:', result.message);
        console.log('Puntuación general:', result.analysis.scores.overall);
        console.log('Transcript:', result.transcript);

        // Mostrar resultados en la UI
        displayResults(result);

    } catch (error) {
        console.error('Error al subir archivo:', error.message);
        alert('Error al subir archivo: ' + error.message);
    }
}

// Función para mostrar resultados en la UI
function displayResults(result) {
    const resultsDiv = document.getElementById('results');
    
    resultsDiv.innerHTML = `
        <h3>Resultados del Análisis</h3>
        <p><strong>Mensaje:</strong> ${result.message}</p>
        <p><strong>Puntuación General:</strong> ${result.analysis.scores.overall}/10</p>
        <h4>Puntuaciones Detalladas:</h4>
        <ul>
            <li>Historia: ${result.analysis.scores.historia}/10</li>
            <li>Misión y Visión: ${result.analysis.scores.mision_vision}/10</li>
            <li>Productos: ${result.analysis.scores.productos}/10</li>
            <li>Valores: ${result.analysis.scores.valores}/10</li>
            <li>Mercado: ${result.analysis.scores.mercado}/10</li>
            <li>Logros: ${result.analysis.scores.logros}/10</li>
        </ul>
        <h4>Feedback:</h4>
        <ul>
            ${result.analysis.feedback.map(feedback => `<li>${feedback}</li>`).join('')}
        </ul>
        <h4>Resumen:</h4>
        <p>${result.analysis.resumen}</p>
        <h4>Transcript:</h4>
        <p>${result.transcript}</p>
    `;
}

// Exportar para uso en módulos
if (typeof module !== 'undefined' && module.exports) {
    module.exports = PresentationAnalysisAPI;
}

// Ejecutar ejemplo si estamos en navegador
if (typeof window !== 'undefined') {
    // Crear elementos de ejemplo en la página
    document.addEventListener('DOMContentLoaded', function() {
        const container = document.createElement('div');
        container.innerHTML = `
            <h2>API de Análisis de Presentaciones</h2>
            <div>
                <input type="file" id="fileInput" accept="video/*">
                <button onclick="uploadFileExample()">Subir y Analizar</button>
            </div>
            <div id="results"></div>
        `;
        document.body.appendChild(container);
    });
} 